packages/client-engine-runtime/src/interpreter/QueryInterpreter.ts

```diff
@@ -218,6 +218,44 @@ export class QueryInterpreter {
         return asList(from).filter((item) => !toSet.has(item))
       }
 
+      case 'distinctBy': {
+        const value = await this.interpretNode(node.args.expr, queryable, scope, generators)
+        const seen = new Set()
+        const result: Value[] = []
+        for (const item of asList(value)) {
+          const key = getRecordKey(item!, node.args.fields)
+          if (!seen.has(key)) {
+            seen.add(key)
+            result.push(item)
+          }
+        }
+        return result
+      }
+
+      case 'paginate': {
+        const value = await this.interpretNode(node.args.expr, queryable, scope, generators)
+        const list = asList(value)
+
+        const linkingFields = node.args.pagination.linkingFields
+        if (linkingFields !== null) {
+          const groupedByParent = new Map<string, Value[]>()
+          for (const item of list) {
+            const parentKey = getRecordKey(item!, linkingFields)
+            if (!groupedByParent.has(parentKey)) {
+              groupedByParent.set(parentKey, [])
+            }
+            groupedByParent.get(parentKey)!.push(item)
+          }
+
+          const groupList: [string, unknown[]][] = Array.from(groupedByParent.entries())
```

Comment:
technically unnecessary type annotation (fine with me if find it helpful for readability though)